pull/github.go

```diff
@@ -309,44 +310,170 @@ func (ghc *GitHubContext) Reviews() ([]*Review, error) {
 	return ghc.reviews, nil
 }
 
-func (ghc *GitHubContext) RepositoryCollaborators() (map[string]string, error) {
+func (ghc *GitHubContext) RepositoryCollaborators() ([]*Collaborator, error) {
 	if ghc.collaborators == nil {
-		opts := github.ListCollaboratorsOptions{
-			ListOptions: github.ListOptions{
-				PerPage: 100,
-			},
+		// For reveiwer assignment, we need to figure out how each collaborator
+		// gets permissions on the repository. We _should_ be able to do this
+		// by examining the permission sources for each collaborator, but this
+		// API is innaccurate as of 2021-05-06. Specifically:
+		//
+		//   - Organization permissions are not reported correctly
+		//   - Triage/Maintain permissions are not supported
+		//   - Organization owners have both org and repo sources
+		//
+		// Instead, query for all collaborators and direct collaborators, then
+		// join that information with team level permissions to produce the
+		// final list of collaborators.
+
+		var q struct {
+			Repository struct {
+				DirectCollaborators struct {
+					PageInfo v4PageInfo
+					Edges    []struct {
+						Permission string
+					}
+					Nodes []v4Actor
+				} `graphql:"direct: collaborators(affiliation: DIRECT, first: 100, after: $directCursor)"`
+				AllCollaborators struct {
+					PageInfo v4PageInfo
+					Edges    []struct {
+						Permission        string
+						PermissionSources []struct {
+							Source struct {
+								Team struct {
+									Slug string
+								} `graphql:"... on Team"`
+							}
+						}
```

Comment:
This _should_ work, and seems to behave correctly with a personal access token, but I can't get `permissionSources` to resolve to a non-empty list when using a GitHub App token. Filed a ticket with support, but will probably have to replace this with direct (and more expensive) team listing to check membership.

Reply:
Replaced with team member listing in the latest commit. Fortunately, I think we'd perform most these listings anyway in any cases where `RepositoryCollaborators` is currently used, so net requests probably aren't increasing too badly.